planner/core/rule_join_reorder_dp.go

```diff
@@ -66,15 +91,37 @@ func (s *joinReorderDPSolver) solve(joinGroup []LogicalPlan, conds []expression.
 			continue
 		}
 		visitID2NodeID := s.bfsGraph(i, visited, adjacents, nodeID2VisitID)
+		nodeIDMask := uint(0)
+		for _, nodeID := range visitID2NodeID {
+			nodeIDMask |= 1 << uint(nodeID)
+		}
+		var subNonEqEdges []joinGroupNonEqEdge
+		for i := len(totalNonEqEdges) - 1; i >= 0; i-- {
+			// If this edge is not the subset of the current sub graph.
+			if totalNonEqEdges[i].idMask&nodeIDMask != totalNonEqEdges[i].idMask {
+				continue
+			}
+			newMask := uint(0)
+			for _, nodeID := range totalNonEqEdges[i].nodeIDs {
+				newMask |= 1 << uint(nodeID2VisitID[nodeID])
+			}
+			totalNonEqEdges[i].idMask = newMask
+			subNonEqEdges = append(subNonEqEdges, totalNonEqEdges[i])
+			totalNonEqEdges = append(totalNonEqEdges[:i], totalNonEqEdges[i+1:]...)
+		}
```

Comment:
Can we put this logic into `nodesAreConnected` to make it consistent with `EqualConditions`?

Reply:
Put it here since we need to update its `idMask` in line 104-108.
If we put it into `nodesAreConnected`. We need to calculate the mask every time, or we need to do the procedure in `dpGraph`'s early stage.
Which one do you prefer?

Reply:
OK, then I prefer leaving it here.